client/utils/src/mpsc.rs

```diff
@@ -16,287 +16,164 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! Features to meter unbounded channels
-
-#[cfg(not(feature = "metered"))]
-mod inner {
-	// just aliased, non performance implications
-	use futures::channel::mpsc::{self, UnboundedReceiver, UnboundedSender};
-	pub type TracingUnboundedSender<T> = UnboundedSender<T>;
-	pub type TracingUnboundedReceiver<T> = UnboundedReceiver<T>;
-
-	/// Alias `mpsc::unbounded`
-	pub fn tracing_unbounded<T>(
-		_key: &'static str,
-	) -> (TracingUnboundedSender<T>, TracingUnboundedReceiver<T>) {
-		mpsc::unbounded()
-	}
+//! Code to meter unbounded channels.
+
+use crate::metrics::UNBOUNDED_CHANNELS_COUNTER;
+use async_channel::{Receiver, Sender, TryRecvError, TrySendError};
+use futures::{
+	stream::{FusedStream, Stream},
+	task::{Context, Poll},
+};
+use log::error;
+use sp_arithmetic::traits::SaturatedConversion;
+use std::{
+	backtrace::Backtrace,
+	pin::Pin,
+	sync::{
+		atomic::{AtomicBool, Ordering},
+		Arc,
+	},
+};
+
+/// Wrapper Type around [`async_channel::Sender`] that increases the global
+/// measure when a message is added.
+#[derive(Debug)]
+pub struct TracingUnboundedSender<T> {
+	inner: Sender<T>,
+	name: &'static str,
+	queue_size_warning: usize,
+	warning_fired: Arc<AtomicBool>,
+	creation_backtrace: Arc<Backtrace>,
```

Comment:
Not a big deal, but technically having *two* `Arc`s here should be unnecessary.

Reply:
Do you mean it's better to use something like `Arc<(AtomicBool, Backtrace)>`?

Reply:
Yeah, or put it in a struct.

But really, it shouldn't really matter; it's just my OCD talking where things which are less efficient than they could be trigger me. Feel free to ignore if you want. :P

Reply:
Okay, this one seems pretty harmless. I'm going to keep two Arcs this time)